lncfg/db.go

```diff
@@ -20,12 +21,21 @@ const (
 	TowerServerDBName = "watchtower.db"
 	WalletDBName      = "wallet.db"
 
+	SqliteChannelDBName  = "channel.sqlite"
+	SqliteChainDBName    = "chain.sqlite"
+	SqliteNeutrinoDBName = "neutrino.sqlite"
+	SqliteTowerDBName    = "watchtower.sqlite"
+
 	BoltBackend                = "bolt"
 	EtcdBackend                = "etcd"
 	PostgresBackend            = "postgres"
+	SqliteBackend              = "sqlite"
 	DefaultBatchCommitInterval = 500 * time.Millisecond
 
 	defaultPostgresMaxConnections = 50
+	defaultSqliteMaxConnections   = 50
+
+	defaultSqliteBusyTimeout = 5 * time.Second
```

Comment:
> Is it good to have this timeout? 

Bolt's behavior when a read transaction needs to upgrade to a write transaction while another write txn is active is simply to wait indefinitely. For sqlite, you need _some_ sort of timeout. Our current architecture here should mostly avoid the need for this, as read transactions are known ahead of time (w/ the kvdb interface), and then write transactions can ensure that they only start once the exclusive lock is available (so no txn upgrade needed). 

If you check out some of the comment history above, many of the timeouts were due to having the wallet data in the same DB as everything else. Today with bolt they're distinct, and we noticed that if we don't do the same with sqlite, then we'd run into it frequently, since at times a db transaction is needed to read out a private key for the first time, etc. 

In terms of the current db handlers, it's also possible to catch this error then try again (implement your own busy loop). In either case, that loop will either need to terminate after a certain amount of times, or never terminate which can lead to livelock situations.

Reply:
>write transactions can ensure that they only start once the exclusive lock is available

I think that even with IMMEDIATE, this can still happen when another write tx runs for longer than the busy timeout? An indefinite timeout seems not very useful, but isn't a five second default too short?

>If you check out some of the comment history above, many of the timeouts were due to having the wallet data in the same DB as everything else.

Definitely aware of that. With postgres we had the same problem.

Reply:
> An indefinite timeout seems not very useful, but isn't a five second default too short?

5 seconds seems pretty long in "computer time", it's what we've set so far and isn't of course final as it'll likely be tuned in the rc phase. We've been using this extensively in some other projects and haven't run into an issue so far. I think the current timeout is nice, as we can find other instances where it's insufficient (which might point to a bug perhaps) and either fix those or bump this. This value is also configurable on the CLI as well.

Reply:
Perhaps it's useful to run https://github.com/bottlepay/lightning-benchmark to really load the nodes and see what it does.

Reply:
I tried this on a standard cloud instance. Results:
bbolt: 100 tps
sqlite (default synchronous=normal): 60 tps
sqlite synchronous=full: `SQLITE_BUSY` ...

Reply:
I did not. But I ran it again with `synchronous=full` and `maxconnections=2` and got the following:

```
lightning-benchmark-loadtest-1   | 2023-01-18T20:00:30.804Z	INFO	Speed	{"tps": 45.59890627331616, "count": 1000, "avg_latency_sec": 1.995817982483}
lightning-benchmark-loadtest-1   | 2023-01-18T20:00:51.593Z	INFO	Speed	{"tps": 48.10364820577079, "count": 2000, "avg_latency_sec": 2.0795732878499997}
lightning-benchmark-loadtest-1   | 2023-01-18T20:01:12.134Z	INFO	Speed	{"tps": 48.68273922414404, "count": 3000, "avg_latency_sec": 2.069190335678}
lightning-benchmark-loadtest-1   | 2023-01-18T20:01:32.449Z	ERROR	Error sending payment	{"err": "payment failed"}
```

Checked the sender lnd log file and saw:

`lightning-benchmark-lnd-alice-1  | 2023-01-18 20:01:32.381 [ERR] HSWC: unable to commit circuit in switch: database is locked (5) (SQLITE_BUSY)`

Reply:
This is the branch: https://github.com/bottlepay/lightning-benchmark/compare/sqlite